File: crates/netcfgd-netlink/src/inotify.rs
//! inotify, for noticing that the config directory changed.
//!
//! Not netlink, and in this crate anyway. Section 1 constraint 4 names
//! `netcfgd-netlink` as the single crate permitted `unsafe`, so what defines
//! it is the audit, not the protocol -- a second crate making raw syscalls
//! would mean a second thing to review to the same bar, which is exactly what
//! the constraint exists to prevent. See `docs/decisions/0012`.
//!
//! The split is the same as for netlink: the syscalls are here with SAFETY
//! comments, and the parsing of what comes back is entirely safe code in
//! [`Events`], with the same termination hazard and the same treatment.

use std::ffi::CString;
use std::io;
use std::os::unix::ffi::OsStrExt;
use std::path::Path;

/// Length of `struct inotify_event` before its trailing name.
pub const EVENT_HDR_LEN: usize = 16;

/// Event masks, from `linux/inotify.h`.
pub mod mask {
	/// A file was written to.
	pub const MODIFY: u32 = 0x0000_0002;
	/// A file open for writing was closed. The usual signal for "an editor
	/// finished saving".
	pub const CLOSE_WRITE: u32 = 0x0000_0008;
	/// Something was moved out of the directory.
	pub const MOVED_FROM: u32 = 0x0000_0040;
	/// Something was moved into it. The other usual editor signal, since a
	/// careful writer renames into place rather than truncating.
	pub const MOVED_TO: u32 = 0x0000_0080;
	/// Something was created in it.
	pub const CREATE: u32 = 0x0000_0100;
	/// Something was deleted from it.
	pub const DELETE: u32 = 0x0000_0200;
	/// The watched directory itself went away.
	pub const DELETE_SELF: u32 = 0x0000_0400;
	/// The watched directory itself was moved.
	pub const MOVE_SELF: u32 = 0x0000_0800;
	/// The kernel's event queue overflowed and events were lost.
	pub const Q_OVERFLOW: u32 = 0x0000_4000;

	/// Everything that means "the config directory changed".
	///
	/// `MODIFY` is included as well as `CLOSE_WRITE` because a writer that
	/// keeps the file open -- `>>` from a script, say -- never produces a
	/// close, and a config that changed without netcfgd noticing is the whole
	/// failure this watch exists to prevent.
	pub const CONFIG: u32 =
		MODIFY | CLOSE_WRITE | MOVED_FROM | MOVED_TO | CREATE | DELETE | DELETE_SELF | MOVE_SELF;
}

/// One event, as read from the descriptor.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Event {
	/// Which watch produced it.
	pub wd: i32,
	/// What happened.
	pub mask: u32,
	/// The name within the watched directory, where there is one.
	pub name: Option<String>,
}

impl Event {
	/// Whether the kernel dropped events before this one.
	///
	/// Handled the same way as netlink's `ENOBUFS`: it means "you missed
	/// something", which for a watcher that re-reads everything is
	/// indistinguishable from an ordinary change.
	#[must_use]
	pub fn overflowed(&self) -> bool {
		self.mask & mask::Q_OVERFLOW != 0
	}
}

/// Walk the events in one read from an inotify descriptor.
///
/// Safe code over bytes the kernel wrote, with the same two obligations as the
/// netlink iterators: it must terminate, and it must not panic. A `len` field
/// large enough to run past the buffer ends iteration rather than indexing out
/// of bounds.
#[derive(Debug, Clone)]
pub struct Events<'a> {
	rest: &'a [u8],
}

impl<'a> Events<'a> {
	/// Start walking `bytes`.
	#[must_use]
	pub fn new(bytes: &'a [u8]) -> Self {
		Self { rest: bytes }
	}
}

impl Iterator for Events<'_> {
	type Item = Event;

	fn next(&mut self) -> Option<Self::Item> {
		if self.rest.len() < EVENT_HDR_LEN {
			return None;
		}
		let wd = i32::from_ne_bytes(self.rest[0..4].try_into().ok()?);
		let mask = u32::from_ne_bytes(self.rest[4..8].try_into().ok()?);
		let len = u32::from_ne_bytes(self.rest[12..16].try_into().ok()?) as usize;

		let total = EVENT_HDR_LEN.checked_add(len)?;
		if total > self.rest.len() {
			self.rest = &[];
			return None;
		}

		let name = if len == 0 {
			None
		} else {
			let raw = &self.rest[EVENT_HDR_LEN..total];
			// The name is NUL-padded to an alignment boundary, so trim at the
			// first NUL rather than trusting `len` to be the string length.
			let end = raw.iter().position(|byte| *byte == 0).unwrap_or(raw.len());
			std::str::from_utf8(&raw[..end]).ok().map(ToOwned::to_owned)
		};

		self.rest = &self.rest[total..];
		Some(Event { wd, mask, name })
	}
}

/// An inotify descriptor.
#[derive(Debug)]
pub struct Inotify {
	fd: libc::c_int,
}

impl Inotify {
	/// Open a descriptor.
	///
	/// # Errors
	///
	/// Returns the underlying `io::Error`. `EMFILE` here means the system's
	/// `fs.inotify.max_user_instances` is exhausted, which is a real and
	/// ordinary condition on a busy machine -- the caller is expected to fall
	/// back rather than fail.
	pub fn new() -> io::Result<Self> {
		// SAFETY: `inotify_init1` takes one integer flag and returns a file
		// descriptor or -1. No pointers are involved, so there is nothing to
		// get wrong about lifetimes or provenance, and -1 is checked.
		let fd = unsafe { libc::inotify_init1(libc::IN_CLOEXEC | libc::IN_NONBLOCK) };
		if fd < 0 {
			return Err(io::Error::last_os_error());
		}
		Ok(Self { fd })
	}

	/// Watch a directory.
	///
	/// # Errors
	///
	/// Returns the underlying `io::Error`, including for a path that does not
	/// exist -- which is not fatal, since a config directory may be created
	/// later.
	pub fn watch(&self, path: &Path, mask: u32) -> io::Result<i32> {
		let raw = CString::new(path.as_os_str().as_bytes())
			.map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "path contains a NUL"))?;
		// SAFETY: `raw` is a live NUL-terminated C string that outlives the
		// call, which is exactly what `inotify_add_watch` requires of its
		// second argument. `self.fd` is valid for as long as `self`.
		let wd = unsafe { libc::inotify_add_watch(self.fd, raw.as_ptr(), mask) };
		if wd < 0 {
			return Err(io::Error::last_os_error());
		}
		Ok(wd)
	}

	/// Wait up to `timeout_ms` for events, and return them.
	///
	/// An empty vector means the wait timed out, which a caller can use as its
	/// own tick.
	///
	/// # Errors
	///
	/// Returns the underlying `io::Error` for anything but a timeout or an
	/// interrupted syscall.
	pub fn wait(&self, timeout_ms: i32) -> io::Result<Vec<Event>> {
		let mut poll = libc::pollfd {
			fd: self.fd,
			events: libc::POLLIN,
			revents: 0,
		};
		// SAFETY: the pointer is to one live, fully initialised `pollfd` that
		// outlives the call, and the count passed is exactly one, so the
		// kernel reads and writes only that struct.
		let ready = unsafe { libc::poll(std::ptr::addr_of_mut!(poll), 1, timeout_ms) };
		if ready < 0 {
			let error = io::Error::last_os_error();
			// A signal during poll is not a failure to watch.
			if error.kind() == io::ErrorKind::Interrupted {
				return Ok(Vec::new());
			}
			return Err(error);
		}
		if ready == 0 {
			return Ok(Vec::new());
		}

		let mut buffer = vec![0_u8; 8192];
		// SAFETY: `buffer` is a live, uniquely borrowed allocation of the
		// length passed, so the kernel writes only within memory we own and
		// may mutate. The descriptor is non-blocking, so this cannot park.
		let read = unsafe {
			libc::read(
				self.fd,
				buffer.as_mut_ptr().cast::<libc::c_void>(),
				buffer.len(),
			)
		};
		if read < 0 {
			let error = io::Error::last_os_error();
			return match error.kind() {
				io::ErrorKind::WouldBlock | io::ErrorKind::Interrupted => Ok(Vec::new()),
				_ => Err(error),
			};
		}
		let read = usize::try_from(read).unwrap_or(0).min(buffer.len());
		Ok(Events::new(&buffer[..read]).collect())
	}
}

impl Drop for Inotify {
	fn drop(&mut self) {
		// SAFETY: `self.fd` was opened by `new`, has not been closed, and is
		// not reachable elsewhere -- `Inotify` is not `Clone` and does not
		// hand the descriptor out. Dropping is therefore the only close.
		unsafe { libc::close(self.fd) };
	}
}

File: c/src/sys/inotify.c
/*
 * inotify.c -- the descriptor, and the walk over what it returns.
 *
 * The syscalls are here and the parsing is safe arithmetic over bytes the
 * kernel wrote, which is the same split `socket.c` and `wire.c` have. The walk
 * carries the same two obligations as the netlink ones -- it terminates, and it
 * does not read past the end -- and watch.h says why they are the same two.
 */
#include "ncfg/watch.h"

#include <errno.h>
#include <poll.h>
#include <string.h>
#include <unistd.h>

/* The header this file skips past is the system's own length, not a second
 * opinion. */
_Static_assert(sizeof(struct inotify_event) == NCFG_INOTIFY_EVENT_HDR_LEN,
    "an inotify event header is sixteen bytes and this system's is not");

int ncfg_inotify_event_overflowed(const ncfg_inotify_event_t *event)
{
	return event && (event->mask & (uint32_t)IN_Q_OVERFLOW) != 0;
}

void ncfg_inotify_walk_start(ncfg_inotify_walk_t *walk, const void *bytes, size_t length)
{
	if (!walk) {
		return;
	}
	walk->rest = (const uint8_t *)bytes;
	walk->remaining = bytes ? length : 0;
}

ncfg_wire_step_t ncfg_inotify_walk_next(ncfg_inotify_walk_t *walk, ncfg_inotify_event_t *out,
    char *err, size_t err_size)
{
	uint32_t claimed = 0;
	int32_t  wd = 0;
	uint32_t mask = 0;
	size_t   total;

	if (!walk || !out) {
		ncfg_error_set(err, err_size, "walking inotify events needs somewhere to go");
		return NCFG_WIRE_BAD;
	}
	memset(out, 0, sizeof(*out));
	if (walk->remaining == 0) {
		return NCFG_WIRE_END;
	}
	if (walk->remaining < NCFG_INOTIFY_EVENT_HDR_LEN) {
		size_t left = walk->remaining;

		/* Exhausted, so a loop that ignores this still terminates. */
		walk->remaining = 0;
		ncfg_error_set(err, err_size,
		    "an inotify event is at least %u bytes and this one is %zu",
		    (unsigned)NCFG_INOTIFY_EVENT_HDR_LEN, left);
		return NCFG_WIRE_BAD;
	}
	/* Native order, by copying the host representation: these are the
	 * kernel's own integers written into a buffer, and a shift-based read
	 * would be right here and wrong on a big-endian machine that nothing
	 * would ever run this on to notice. */
	memcpy(&wd, walk->rest, sizeof(wd));
	memcpy(&mask, walk->rest + 4, sizeof(mask));
	/* Bytes 8 to 11 are the cookie, which pairs a `MOVED_FROM` with its
	 * `MOVED_TO`. Nothing here needs it: this watcher reports that something
	 * moved and the caller re-reads everything. */
	memcpy(&claimed, walk->rest + 12, sizeof(claimed));

	total = NCFG_INOTIFY_EVENT_HDR_LEN + (size_t)claimed;
	if (total < NCFG_INOTIFY_EVENT_HDR_LEN || total > walk->remaining) {
		walk->remaining = 0;
		ncfg_error_set(err, err_size,
		    "an inotify event claims %u bytes of name and there is not room for it",
		    (unsigned)claimed);
		return NCFG_WIRE_BAD;
	}

	out->wd = wd;
	out->mask = mask;
	if (claimed > 0) {
		const uint8_t *raw = walk->rest + NCFG_INOTIFY_EVENT_HDR_LEN;
		size_t         end = 0;

		/* The name is NUL-padded to an alignment boundary, so it ends at
		 * the first NUL rather than at `len` -- which is the padded
		 * length and not the string's. */
		while (end < (size_t)claimed && raw[end] != '\0') {
			end++;
		}
		if (end < sizeof(out->name)) {
			memcpy(out->name, raw, end);
			out->name[end] = '\0';
			out->has_name = 1;
		}
		/* A name that would not fit is absent rather than truncated: a
		 * truncated name is a name, just somebody else's -- and this one
		 * is used to decide which file changed. `NAME_MAX` says the
		 * kernel cannot produce one, so this is the unreachable arm
		 * being made harmless rather than a case being handled. */
	}
	walk->rest += total;
	walk->remaining -= total;
	return NCFG_WIRE_OK;
}

void ncfg_inotify_init(ncfg_inotify_t *inotify)
{
	if (!inotify) {
		return;
	}
	inotify->fd = -1;
}

int ncfg_inotify_open(ncfg_inotify_t *inotify, char *err, size_t err_size)
{
	int opened;

	if (!inotify) {
		ncfg_error_set(err, err_size, "an inotify descriptor needs somewhere to go");
		return 0;
	}
	ncfg_inotify_init(inotify);
	/* Non-blocking, so the read after a `poll` cannot park: a watcher that
	 * blocked in `read` would be a watcher whose timeout does not
	 * exist. Close-on-exec because netcfgd spawns children. */
	opened = inotify_init1(IN_CLOEXEC | IN_NONBLOCK);
	if (opened < 0) {
		ncfg_error_set(err, err_size, "cannot watch for configuration changes: %s%s",
		    strerror(errno),
		    errno == EMFILE ? "; fs.inotify.max_user_instances is exhausted" : "");
		return 0;
	}
	inotify->fd = opened;
	return 1;
}

void ncfg_inotify_close(ncfg_inotify_t *inotify)
{
	if (!inotify || inotify->fd < 0) {
		return;
	}
	(void)close(inotify->fd);
	inotify->fd = -1;
}

int ncfg_inotify_watch(const ncfg_inotify_t *inotify, const char *path, uint32_t mask,
    int *wd, char *err, size_t err_size)
{
	int taken;

	if (wd) {
		*wd = -1;
	}
	if (!inotify || inotify->fd < 0 || !path) {
		ncfg_error_set(err, err_size, "watching a directory needs an open descriptor");
		return 0;
	}
	taken = inotify_add_watch(inotify->fd, path, mask);
	if (taken < 0) {
		ncfg_error_set(err, err_size, "cannot watch %s: %s", path, strerror(errno));
		return 0;
	}
	if (wd) {
		*wd = taken;
	}
	return 1;
}

int ncfg_inotify_wait(const ncfg_inotify_t *inotify, int timeout_ms,
    ncfg_inotify_batch_t *out, char *err, size_t err_size)
{
	struct pollfd waiting;
	ssize_t       taken;
	int           ready;

	if (!out) {
		ncfg_error_set(err, err_size, "waiting for events needs somewhere to put them");
		return 0;
	}
	out->length = 0;
	if (!inotify || inotify->fd < 0) {
		ncfg_error_set(err, err_size, "waiting for events needs an open descriptor");
		return 0;
	}
	waiting.fd = inotify->fd;
	waiting.events = POLLIN;
	waiting.revents = 0;
	ready = poll(&waiting, 1, timeout_ms);
	if (ready < 0) {
		/* A signal during the wait is not a failure to watch. Reported
		 * as a timeout, which is what the caller's loop already handles
		 * -- and it cannot spin, because every one of them is a signal
		 * that really arrived. */
		if (errno == EINTR) {
			return 1;
		}
		ncfg_error_set(err, err_size, "waiting for configuration changes: %s",
		    strerror(errno));
		return 0;
	}
	if (ready == 0) {
		return 1;
	}
	taken = read(inotify->fd, out->bytes, sizeof(out->bytes));
	if (taken < 0) {
		/* Nothing there after all, or a signal: both are "no events",
		 * which is what a timeout is. */
		if (errno == EAGAIN || errno == EWOULDBLOCK || errno == EINTR) {
			return 1;
		}
		ncfg_error_set(err, err_size, "reading configuration changes: %s",
		    strerror(errno));
		return 0;
	}
	out->length = (size_t)taken;
	return 1;
}
